### PR TITLE
Color Controls: Merge color dropdown components

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related:
- https://github.com/WordPress/gutenberg/pull/34027

## What?
In https://github.com/WordPress/gutenberg/pull/34027, a new `ToolsPanelColorDropdown` component was introduced to facilitate color dropdowns being rendered within the `ToolsPanel` used by the color block supports slot. 

As a result of further [discussion](https://github.com/WordPress/gutenberg/pull/34027#discussion_r841646722), this PR looks to merge this `ToolsPanelColorDropdown` with the original `ColorGradientSettingsDropdown`. 

## Why?

Given the overlap, refactoring these to a single component, should reduce maintenance, highlight how color controls may need to be rendered differently in different contexts, and help ensure both cases are updated appropriately should the color control need changing.

## How?

- Updates the original `ColorGradientSettingsDropdown` component
- Adds a new prop to toggle between rendering as an `ItemGroup` (original component) or as a `ToolsPanelItem`
- Replaces use of `ToolsPanelColorDropdown` with the new merged `ColorGradientSettingsDropdown` component for the color block supports
- Relocates the color dropdown styles from the color block supports file to live with the original component styles in `packages/block-editor/src/components/colors-gradients/style.scss`

## Testing Instructions

1. Load the block editor, create a new post and add both a group and cover block
2. Select the group block and ensure that all the color controls in the sidebar still function and look correct
    (This tests the `ToolsPanelItem` version of the color dropdown component)
3. Make sure to test resetting the color controls via the `ToolsPanel` menu
4. Select the cover block ensuring that the color control in the Overlay sidebar panel is looking and working correctly.
    (This tests the original `ItemGroup` version of the color dropdown component)
5. Save the post and double-check the frontend applies color selections.

## Screenshots or screencast 

https://user-images.githubusercontent.com/60436221/161907959-82a9cd5e-dd5f-480d-b506-15c91d56d0e3.mp4

### CODE REVIEW

File: packages/block-editor/src/components/colors-gradients/dropdown.js
Comment (andrewserong) [inline packages/block-editor/src/components/colors-gradients/dropdown.js]: Not a blocker, but now that these two components are combined into one, would it be worth matching the `__` in the classnames here? Perhaps something like:

```
block-editor-tools-panel-color-gradient-settings__dropdown
```

That way it's similar to the class name in `WithToolsPanelItem`, too.

File: packages/block-editor/src/components/colors-gradients/dropdown.js
Comment (andrewserong) [inline packages/block-editor/src/components/colors-gradients/dropdown.js]: You've already flagged it for a follow-up, but it looks like `renderToggleItem` and `renderToggleButton` could potentially be combined, but for the moment keeping them separate looks fine to me!

**Review by andrewserong — APPROVED** [review]
Thanks for following up with this one @aaronrobertshaw! It took me a little while to wrap my head around the value of the combined component, because it does introduce some additional complexity, where I found the two separate components a little easier to read in isolation.

But, I think the benefits of having a single component as the entry point to rendering the controls makes it worthwhile, and it looks like there's some further consolidation we can do to neaten things up, too, in follow-ups should we wish to do that.

It's also good to see the styles moved over from `hooks/color.scss` to the `colors-gradients` directory 👍

This LGTM, but you might want a second opinion from @youknowriad since he suggested the idea (or otherwise @ciampo always has good insights into these things!)

File: packages/block-editor/src/components/colors-gradients/dropdown.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/dropdown.js]: Thanks, I had similar thoughts so had added that under the possible next steps in the PR description. Given the soft endorsement of combining the components, I'll update this and toggle rendering while awaiting further review.

File: packages/block-editor/src/components/colors-gradients/dropdown.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/dropdown.js]: Thanks, I should have some time today where I can clean this up while the PR awaits extra eyes.

**Comment by aaronrobertshaw** [comment]
Thanks for the review @andrewserong 👍 

I think you capture the tradeoff well in your comment above. Given we appear to be leaning towards this PR's direction I've made the suggested updates to keep this moving.

**Comment by andrewserong** [comment]
Thanks @aaronrobertshaw, from a quick skim that's looking good to me, and makes things a bit tidier. Also, on balance, this PR removes more than it adds, which is a win! 👍

image

**Comment by youknowriad** [comment]
I'd love a quick look from @jorgefilipecosta as he was involved in the initial components here quiet a lot.

File: packages/block-editor/src/components/colors-gradients/dropdown.js
Comment (ciampo) [inline packages/block-editor/src/components/colors-gradients/dropdown.js:15]: It doesn't seem like the `<Item>` component is used at the moment — this may be a bit of a "semantics" regression, since `ItemGroup` renders a element with `role="list"` which would normally need children with `role="listitem"`, which is what essentially the `Item` component does (the `<li>` element has implicitly that role too).

Given that we discussed previously that using `Item` can be limiting (especially when it needs to be wrapped in a `Dropdown` or `Popover`), we should first see if the current state of this PR is semantically good (e.g. the browser effectively parses children of `ItemGroup` as list items).

If not, we have afew choices:
- try to use `Item` (although, as explained above, it may not be possible)
- try to apply `role="listitem"` manually on direct children of `ItemGroup` (and see if that works)
- otherwise, we may also decide to "avoid" the list semantics altogether, and override the `role` attribute on the `ItemGroup` component so that it's not marked as a list anymore

What do folks think?

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/dropdown.js:15]: > It doesn't seem like the `<Item>` component is used at the moment

I'm not sure I follow sorry, the `Item` component is [used for the dropdown toggle](https://github.com/WordPress/gutenberg/pull/40084/files/142ae741392fe7a3dba80bea5baa69521189e652#diff-1d8706a429c020115759594b52150542acec1b47a901b4c09163f3a19b78f9faR79).

> this may be a bit of a "semantics" regression, since ItemGroup renders a element with role="list" which would normally need children with role="listitem", which is what essentially the Item component does

This component (with the original `ItemGroup` > `Item` rendering) is still in use by `PanelColorGradientSettings`. This can be seen in action via the Cover block's overlay color control. e.g. Step 4 in the PR test instructions.

If used via `PanelColorGradientSettings` or with the `isItemGroup` set to `true`, we'll get something akin to the following:

```
<ItemGroup>
  <Item />
  <Item />
  <Item />
</ItemGroup>
```

If `isItemGroup` is set to `false`, such as for the color block supports which need to render the dropdowns within a `ToolsPanel` where we cannot use an `ItemGroup` due to possibly hidden controls, we will get something like:

```
<ToolsPanelItem />
<ToolsPanelItem />
<ToolsPanelItem />
```


> What do folks think?

If my understanding of your comment is correct, this discussion might illustrate the concerns that merging these two very similar color dropdown components would cause confusion due to the increased complexity, even if there is only a single entry point. 

One option to reduce potential confusion might be renaming the `WithItemGroup` and `WithToolsPanelItem` components. They were named that way to match naming where conditional wrapping had occurred elsewhere in the codebase.

Comment (ciampo) [reply packages/block-editor/src/components/colors-gradients/dropdown.js:15]: > I'm not sure I follow sorry, the Item component is used for the dropdown toggle.

That's my bad — I had missed that ternary during my previous review.

> If used via `PanelColorGradientSettings` or with the `isItemGroup` set to `true`, we'll get something akin to the following:
>
> ```
> <ItemGroup>
>  <Item />
>  <Item />
>  <Item />
> </ItemGroup>
> ```

Thank you for the clarification — in this case, then the components tree looks fine!

> If my understanding of your comment is correct, this discussion might illustrate the concerns that merging these two very similar color dropdown components would cause confusion due to the increased complexity, even if there is only a single entry point. 

There is definitely complexity in this code, and I definitely got a bit confused when writing my previous comment (sorry again for my misunderstandings!)

I'm not sure if there's a way to "reduce" this complexity — maybe we could keep the components with their current names, but add more inline comments to explain how the "components tree" changes based on the value of `isItemGroup` ?

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/dropdown.js:15]: I've added some comments in https://github.com/WordPress/gutenberg/pull/40084/commits/4efe4b48ff6bfd6c164108a4d685e63a92b7907e let me know if you think that is sufficient.

Comment (andrewserong) [reply packages/block-editor/src/components/colors-gradients/dropdown.js:15]: Those additional comments read well to me, I think it makes a big difference for the readability of the file. Nice!

Comment (ciampo) [reply packages/block-editor/src/components/colors-gradients/dropdown.js:15]: Much better, thank you! It should definitely help 😄

**Comment by aaronrobertshaw** [comment]
Thank you everyone for the feedback and suggestions. I believe we have a consensus to proceed with this refactor. 

@jorgefilipecosta Do you have any objections with this PR landing?

File: packages/block-editor/src/hooks/color-panel.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/hooks/color-panel.js]: Do we really need the isItemGroup prop can we just use the new UI by default everywhere? Recently we changed the component to show the palettes in a popover everywhere without a prop and that was an even a bigger change. I guess we can just use tools panels everywhere.

**Review by jorgefilipecosta — COMMENTED** [review]
This PR looks good, nice change @aaronrobertshaw 👍 I'm not sure if we need an optional property to allow the component to be used in two versions. I think we can just use the tools panel everywhere, so all the blocks using colors automatically get the new UI.

Currently, the cover, social links, etc, are not using the new UI. These may get solved if we address the previous point. If not we should enable the new flag on these blocks. If we really introduce the flag I think it should be experimental.

**Comment by aaronrobertshaw** [comment]
Thanks for taking the time to look at this @jorgefilipecosta 👍  

Apologies in advance for the long reply.

> I'm not sure if we need an optional property to allow the component to be used in two versions. I think we can just use the tools panel everywhere, so all the blocks using colors automatically get the new UI.

I'm not sure whether we can just use the ToolsPanel everywhere. 

The `ToolsPanel` requires additional props and callbacks in order to function correctly and this new component version was more about keeping the same UI when within a `ToolsPanel` context. It's also web only.

Using the tools panel everywhere would also likely entail removing or updating the `PanelColorGradientSettings` and `PanelColorSettings` components. That might be easier for the first given it is exported as `__experimental` (though it's been around over 2 years). The `PanelColorSettings` however is a stable component and switching it to a `ToolsPanel` would be a breaking change.

Furthermore, I can still see some blocks wishing to add arbitrary color controls in a semantic way outside of a `ToolsPanel` though. For example, the Cover block has overlay and opacity controls. The overlay control would fit nicely within the color support tools panel however the opacity looks a little out of place. If this were a 3rd party block, the author might wish to keep these two related controls together in a separate panel.

Screen Shot 2022-04-27 at 3 37 36 pm


> Currently, the cover, social links, etc, are not using the new UI.

We can, as you suggest, update the Cover and Social Links blocks to inject their color-related controls into the color block supports `ToolsPanel`.  The Navigation block also uses `PanelColorSettings` which currently leverages the original version of the color dropdown controls. These updates would reduce or remove the use of `PanelColorGradientSettings` and `PanelColorSettings` within Gutenberg but still leaves the question of continuing to support and not breaking  `PanelColorSettings`.

> Recently we changed the component to show the palettes in a popover everywhere without a prop and that was an even a bigger change.

The difference I see here compared to updating the component to show the palette in a popover is, at that point in time there was only one type of panel the color controls were used in so they all had the same constraints. Nothing really changed for their consumers.

The conversion of the [colors block support to use a tools panel](https://github.com/WordPress/gutenberg/pull/34027) was pending review and approval when the [popover updates](https://github.com/WordPress/gutenberg/pull/37067) to the color controls were made. I was under the impression there was a use case for rendering them in an `ItemGroup`. Combining that with the issues around `PanelColorSettings` etc. it was decided to avoid replacing the color dropdown component but create a version that worked alongside it.

---

I don't feel strongly about which way we go, I'd like to clean up the code here and reduce the complexity as much as possible while we are at it.

So that said, some options to move forward could be to:

1. Proceed with this PR
	- If so, possibly make the `isItemGroup` prop `__experimental` (the whole component is currently experimental)
	- Optionally, update Cover, Social Links & Navigation blocks to inject controls into color supports panel anyway
2. Remove non-ToolsPanel use of color controls and find a means of deprecating or removing `PanelColorSettings` and `PanelColorGradientSettings` etc.
	- Will also require fixing native controls where `ToolsPanel` isn't a thing.
3. Abandon this PR
    - Possibly move the separate `ToolsPanelColorDropdown` to an independent location in the block editor components
    - This fails to address the original concern where there were essentially two locations to update if the color controls UI need to change.
    
Thanks again for your help and guidance on this one @jorgefilipecosta 🙇

**Review by jorgefilipecosta — APPROVED** [review]
Thank you for the detailed explanation @aaronrobertshaw.

I think this PR is a step in the good direction of unifying color components. For now, I think it would be better to have isItemGroup as experimental so that in the future, we are able to remove it.

In the future, I think it would be good to try to make PanelColorGradientSettings and PanelColorSettings use the tools panel, but we can try that in different PR's as we are not sure if it fits or not. I think we can keep the same props the components have. And by default, if we provide this experience in every block it is a good idea, on the cover the default may not be optimal but it is not too bad either following the screen @aaronrobertshaw shared.
165460066-d46d842f-ae38-4256-a57b-9826b51af6ef (1)


Regarding adding additional props to PanelColorGradientSettings to make it fit a tools panel I don't think we need that the PanelColorGradientSettings already can reset and manipulate the colors it would be a matter of inside it adapting the properties so a tools panel could be used.

**Comment by aaronrobertshaw** [comment]
Thanks for giving this another look @jorgefilipecosta 🙇 

> I think this PR is a step in the good direction of unifying color components. For now, I think it would be better to have isItemGroup as experimental so that in the future, we are able to remove it.

✅ Done. I've made `isItemGroup` experimental in https://github.com/WordPress/gutenberg/pull/40084/commits/025392fc9fd23db5cf68b820f963bf486d00b3ff.

> In the future, I think it would be good to try to make PanelColorGradientSettings and PanelColorSettings use the tools panel, but we can try that in different PR's as we are not sure if it fits or not.

I agree it would be nice to unify everything here to use the `ToolsPanel` and explore that all separately.

> Regarding adding additional props to PanelColorGradientSettings to make it fit a tools panel I don't think we need that the PanelColorGradientSettings already can reset and manipulate the colors it would be a matter of inside it adapting the properties so a tools panel could be used.

When I mentioned the `ToolsPanel` requiring additional props and callbacks, I meant that for each color control to be wrapped in a `ToolsPanelItem`, the settings object representing the color control should have additional properties including; `hasValue`, `onDeselect`, and `resetAllFilter`. We had a similar requirement for the [color block supports](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/color.js#L520-L582).

These are required for the panel to behave as per the others. The color control itself might allow for a value to be cleared but it will still be represented within the `ToolsPanel` menu. From there it should be able to be reset either individually (`onDeselect`) or via the "reset all" (`resetAllFilter`). The `ToolsPanel` would also need a callback (`hasValue`) to determine whether any selection has been made via the control so as to reflect that in the menu.

We could easily do all this for our uses of the `PanelColorGradientSettings` and `PanelColorSettings` however given `PanelColorSettings` has been exported as stable we'd still need third party devs to update their usage or the resulting panel would have a confusing/broken UX.

For now, I believe we have a consensus to land this PR and explore the rest separately. I'll aim to merge this tomorrow pending any last minute objections.

**Comment by jorgefilipecosta** [comment]
Yes, the consensus is to merge this PR.

> When I mentioned the ToolsPanel requiring additional props and callbacks, I meant that for each color control to be wrapped in a ToolsPanelItem, the settings object representing the color control should have additional properties including; hasValue, onDeselect, and resetAllFilter. We had a similar requirement for the [color block supports](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/color.js?rgh-link-date=2022-05-11T05%3A02%3A32Z#L520-L582).


PanelColorSettings may create all the props (hasValue, onDeselect, and resetAllFilter) based just on the color values before passing them to the ToolsPanel. So I think we can make PanelColorSettings use ToolsPanel without developers needing to pass any new prop to PanelColorSettings but that's something we can try in the future.

**Comment by aaronrobertshaw** [comment]
Thanks everyone for the feedback and help getting this one merged. 🙇 

> PanelColorSettings may create all the props (hasValue, onDeselect, and resetAllFilter) based just on the color values before passing them to the ToolsPanel

I'm a bit at a loss as to how we'd be able to reliably create the `resetAllFilter`. Some colors could be used ad hoc, others from block attributes, they may or may not have named color block attributes as well that should be cleared etc. Making the panel's `resetAll` chain on change handlers would also hit problems. 

If you have some ideas here, I'd be happy to help when the time comes to explore them.

**Pull request merged**
